### PR TITLE
[DOC]: Add docstring example to the contributing docs

### PR DESCRIPTION
**Pull request by echedey-ls**
- [x] Partially addresses #2081
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

Adds a docstring example to the contributing page. Now that I came back to the conversation in #2081 I see it was a discussion between @IoannisSifnaios and @AdamRJensen , so sorry for getting into it.

I still hesitate a bit on whether this is the preferred format to include this template here, maybe just linking to well documented functions is enough.

### CODE REVIEW

**Comment by cwhanse** [comment]
What do we think about these directives? https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#describing-changes-between-versions

I rather like them but don't want to introduce something that's a pain to manage. If we like them, it would be nice to show one in the example.

File: docs/sphinx/source/contributing/style_guide.rst [hunk 1]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        Parameters
+        ----------
+        poa_global : numeric
+            Plane-of-array global irradiance, see :term:`poa_global`. [Wm⁻²].
+        exponents : array-like
+            A list of exponents. [x⁰¹²³⁴⁵⁶⁷⁸⁹⁺⁻].
+        degree_symbol : numeric
+            It's different from superscript zero. [°].
+        default_arg : string, default ``'someval'``.
+            Text enclosed in double backticks mean code formatting.
```
Comment (AdamRJensen) [inline docs/sphinx/source/contributing/style_guide.rst:146]:
```suggestion
        default_arg_1 : string, default ``'someval'``.
            Text enclosed in double backticks mean code formatting.
        time_ref : string, {'UT', 'TST'}, default: 'UT'
            'UT' (universal time) or 'TST' (True Solar Time)
```
How about adding one where there is a set of values you need to choose from?

File: docs/sphinx/source/contributing/style_guide.rst [hunk 2]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+                        default_arg='someval', optional_arg=None):
+        r"""
+        A brief summary of the function.
+
+        A longer description of the function. This can include references
+        to literature [1]_, websites, and other code elements such as
+        functions (:py:func:`pvlib.location.lookup_altitude`) and
+        classes (:py:class:`pvlib.location.Location`).
+
+        .. versionadded:: 0.0.1
```
Comment (AdamRJensen) [inline docs/sphinx/source/contributing/style_guide.rst:131]: Can we add a sentence/[link](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#describing-changes-between-versions) here and maybe also namedrop the other options, e.g., versionchanged, deprecated, and versionremoved

File: docs/sphinx/source/contributing/style_guide.rst (hunk 1 above)
Comment (echedey-ls) [reply docs/sphinx/source/contributing/style_guide.rst:146]: That has never been a standard in pvlib, right? I've usually found the possible values listed in the param description body. Don't get me wrong, I'd love pvlib to have that + type annotations with the possible values. My only remarks are the length when there is a longer list of items and the colon, I find `:` is visual noise.

```suggestion
        default_arg_1 : string, default ``'someval'``.
            Text enclosed in double backticks mean code formatting.
        time_ref : ``'UT'`` or ``'TST'``, default ``'UT'``
            ``'UT'`` (universal time) or ``'TST'`` (True Solar Time)
```

Comment (kandersolar) [reply docs/sphinx/source/contributing/style_guide.rst:146]: Maybe not a standard, but it is used:

```python
pvlib-python/pvlib$ git grep ": {.*}, default"
iotools/pvgis.py:    pvtechchoice: {'crystSi', 'CIS', 'CdTe', 'Unknown'}, default: 'crystSi'
iotools/pvgis.py:    mountingplace: {'free', 'building'}, default: free
iotools/pvgis.py:    trackingtype: {0, 1, 2, 3, 4, 5}, default: 0
iotools/solaranywhere.py:    time_resolution: {60, 30, 15, 5}, default: 60
iotools/solaranywhere.py:    spatial_resolution: {0.1, 0.01, 0.005}, default: 0.01
iotools/solaranywhere.py:    missing_data: {'Omit', 'FillAverage'}, default: 'FillAverage'
iotools/solargis.py:    timestamp_type : {'start', 'center', 'end'}, default: 'center'
pvsystem.py:    fill_method : {'ffill', 'bfill', 'nearest'}, default 'ffill'
```

File: docs/sphinx/source/contributing/style_guide.rst [hunk 3]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
 
 
+.. _example-docstring:
+
+Example Docstring
+-----------------
+
+Here is a template for a function docstring that tries to encapsulate all
+the features that may be used in any docstring. This is a template, so
+not all sections are required for every function, but is a good starting point.
```
Comment (RDaxini) [inline docs/sphinx/source/contributing/style_guide.rst:117]: This sounds more concise, but is just a suggestion
```suggestion
Here is a template for a function docstring that encapsulates the main
features that may be used in any docstring. Note that not all sections are
required for every function.
```

File: docs/sphinx/source/contributing/style_guide.rst [hunk 4]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+
+.. code-block:: python
+
+    def example_function(poa_global, exponents, degree_symbol,
+                        default_arg='someval', optional_arg=None):
+        r"""
+        A brief summary of the function.
```
[the remark this answers is not in the document]
Comment (RDaxini) [reply docs/sphinx/source/contributing/style_guide.rst:124]: Good point 👍🏾
I lean towards omitting the "though" ...though
Sounds a bit casual/seems unnecessary?

File: docs/sphinx/source/contributing/style_guide.rst [hunk 5]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        -----
+        This section can include additional information about the function.
+
+        For example, an equation using LaTeX markup:
+
+        .. math::
+
+            a = \left(\frac{b}{c}\right)^2
+
+        where :math:`a` is the result of the equation, :math:`b` and :math:`c`
```
Comment (RDaxini) [inline docs/sphinx/source/contributing/style_guide.rst:172]:
```suggestion
        where :math:`a` is the result of the equation, and :math:`b` and :math:`c`
```

File: docs/sphinx/source/contributing/style_guide.rst [hunk 6]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        are inputs.
+
+        Or a figure with a caption:
```
Comment (RDaxini) [inline docs/sphinx/source/contributing/style_guide.rst:175]: Is it worth adding somewhere where such figures should be stored in the pvlib directory?

File: docs/sphinx/source/contributing/style_guide.rst [hunk 7]
```diff
@@ -107,6 +107,107 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        A IEEE citation to a relevant reference. You may use an automatic
+        citation generator to format the citation correctly.
+
+        .. [1] Anderson, K., Hansen, C., Holmgren, W., Jensen, A., Mikofski, M.,
+        and Driesse, A. “pvlib python: 2023 project update.” Journal of Open
+        Source Software, 8(92), 5994, (2023). :doi:`10.21105/joss.05994`.
+        """
+        return "None"
+
+A preview on how this would render in the documentation can be seen in the
```
Comment (RDaxini) [inline docs/sphinx/source/contributing/style_guide.rst:205]:
```suggestion
A preview of how this docstring would render in the documentation can be seen in the
```

**Review by RDaxini — COMMENTED** [review]
Nice job. I think this example will certainly help contributors in the future

**Review by IoannisSifnaios — COMMENTED** [review]
Personally this is exactly what I had in mind when we were discussing this! Nice work @echedey-ls! Some suggestions from my side.

File: docs/sphinx/source/contributing/style_guide.rst [hunk 8]
```diff
@@ -107,6 +107,110 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        A IEEE citation to a relevant reference. You may use an automatic
+        citation generator to format the citation correctly.
+
+        .. [1] Anderson, K., Hansen, C., Holmgren, W., Jensen, A., Mikofski, M.,
+           and Driesse, A. "pvlib python: 2023 project update." Journal of Open
+           Source Software, 8(92), 5994, (2023). :doi:`10.21105/joss.05994`.
+        .. [2] J. Smith and J. Doe. "Obama inaugurated as President." CNN.com.
+           Accessed: Feb. 1, 2009. [Online.]
+           Available: http://www.cnn.com/POLITICS/01/21/obama_inaugurated/index.html
+        """
```
[the remark this answers is not in the document]
Comment (echedey-ls) [reply docs/sphinx/source/contributing/style_guide.rst:205]: I think I don't understand you. Print statements are discouraged in the contributing guidelines ([source](https://github.com/pvlib/pvlib-python/blob/c52e600b996733a4a3371623a3784ccda5667e68/docs/sphinx/source/contributing/style_guide.rst?plain=1#L28C1-L30C1)) so I wouldn't add any of them. What currently gets printed in the docstring example is the result if it gets run in an interactive shell.

File: docs/sphinx/source/contributing/style_guide.rst (hunk 6 above)
Comment (echedey-ls) [reply docs/sphinx/source/contributing/style_guide.rst:180]: I think that is intuitive enough if you see the repo layout and that most of the time us contributors copy&paste from other places. Feel free to weight in 👀

File: docs/sphinx/source/contributing/style_guide.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (AdamRJensen) [reply docs/sphinx/source/contributing/style_guide.rst:209]: I think Ioannis was just missing one line of code.

I agree with @echedey-ls though, that it shouldn't be a print statement. I've left a simple suggestion.

File: docs/sphinx/source/contributing/style_guide.rst [hunk 9]
```diff
@@ -107,6 +107,114 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        return "Something"
```
Comment (AdamRJensen) [inline docs/sphinx/source/contributing/style_guide.rst:210]:
```suggestion
        a = 'some'
        b = 'thing'
        return a + b
```

Comment (AdamRJensen) [reply docs/sphinx/source/contributing/style_guide.rst:210]: I think it was something like this @IoannisSifnaios was looking for.

File: docs/sphinx/source/contributing/style_guide.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (IoannisSifnaios) [reply docs/sphinx/source/contributing/style_guide.rst:209]: Yes, that is what I meant (though I admit I was not aware that we discourage print statements, so thanks for that). I am fine with the code @AdamRJensen added below!

File: docs/sphinx/source/contributing/style_guide.rst [hunk 10]
```diff
@@ -107,6 +107,116 @@ Note that Windows users need not have the ``make`` utility installed as pvlib
...
+        """
+        a = "Some"
+        b = "thing"
+        return a + b
+
+A preview of how this docstring would render in the documentation can be seen in the
+following file: :download:`Example docstring<../_images/example_function_screenshot.png>`.
+
+Remember that to show the docstring in the documentation, you must list
+the function in the appropriate ``.rst`` file in the ``docs/sphinx/source/reference`` file.
```
Comment (cwhanse) [inline docs/sphinx/source/contributing/style_guide.rst:218]:
```suggestion
the function in the appropriate ``.rst`` file in the ``docs/sphinx/source/reference`` folder.
```

**Review by kandersolar — APPROVED** [review]
This looks great to me.  I bet an example like this will be a good reference not just for pvlib, but other packages too, and even people's personal code.

**Comment by kandersolar** [comment]
Thanks @echedey-ls and all reviewers :)

**Pull request merged**